orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/servergroup/ServerGroupCacheForceRefreshTask.groovy

```diff
@@ -154,26 +157,46 @@ class ServerGroupCacheForceRefreshTask extends AbstractCloudProviderAwareTask im
                                                   String cloudProvider,
                                                   StageData stageData,
                                                   Long startTime) {
+    // log.debug("Force cache refresh calling clouddriver with refresh type ${REFRESH_TYPE}")
     def pendingForceCacheUpdates = cacheStatusService.pendingForceCacheUpdates(cloudProvider, REFRESH_TYPE)
-    boolean isRecent = (startTime != null) ? pendingForceCacheUpdates.find { it.cacheTime >= startTime } : false
+    log.debug("Force cache refresh clouddriver response was: ${pendingForceCacheUpdates}.")
+    boolean pendingFound = pendingForceCacheUpdates.find { it.cacheTime >= startTime }
+    boolean isRecent = (startTime != null) ? (pendingFound != null ) ? pendingFound : false : false
+
+    log.debug("Force cache refresh isRecent = ${isRecent}")
 
     boolean finishedProcessing = true
     stageData.deployServerGroups.each { String region, Set<String> serverGroups ->
-      def makeModel = { serverGroup ->
-        [serverGroup: serverGroup, region: region, account: account]
-      }
+      def makeModel = { serverGroup -> [serverGroup: serverGroup, region: region, account: account] }
 
       def processedServerGroups = serverGroups.findAll { String serverGroup ->
         def model = makeModel(serverGroup)
 
+        def forceCacheUpdate = pendingForceCacheUpdates.find {
+          (it.details as Map<String, String>).intersect(model) == model
+        }
+
+        if (!stageData.processedServerGroups.contains(model) && forceCacheUpdate == null) {
+          log.debug("Force cache refresh response narrowed to ${model} was null, and this group hasn't been processed yet!")
+        }
+
         if (stageData.processedServerGroups.contains(model)) {
           // this server group has already been processed
+          log.debug("Force cache refresh the model ${model} has been already processed, returning true.")
           return true
         }
 
+        if (!forceCacheUpdate) {
+          // there is no pending cache update, force it again in the event that it was missed
+          stageData.removeRefreshedServerGroup(model.serverGroup, model.region, model.account)
+          log.warn("Unable to find pending cache refresh request (model: ${model}), forcing a new cache refresh.")
+
+          return false
```

Comment:
I'm confused what the change is here -- the `if (!forceCacheUpdate) {` and `def forceCacheUpdate = ...` chunks of code were just copied from below. The only thing that seems to happen differently is that we return `false` in this case earlier than we otherwise would have anyway

Reply:
Earlier, isRecent would be null and stageData.removeRefreshedServerGroup wouldn't be called, thus orca wouldn't POST another ondemand request to clouddriver. It would keep GET-ing cache refresh data from clouddriver, setting isRecent to null and go over and over. Other option would be adding to the `if (!isRecent)` condition, but I don't like duplicating code.

Reply:
In other words, checking for isRecent or not should be only done when we get a valid response from clouddriver (a response with the servergroup we are interested in). So I am fixing the logic here. If response did not have data with the server group we need, we assume that clouddriver lost it somewhere and we put a request for it in the next cycle.

Agreed?

Reply:
Ah got it - yes that sounds better. Can @ajordens or @cfieber double check?